### PR TITLE
user wallet linking for bitflier

### PR DESCRIPTION
**Pull request by husobee**
### Summary

Adds in user wallet linking for bitflyer integration.

### Type of change ( select one )

- [x] Product feature
- [ ] Bug fix
- [ ] Performance improvement
- [ ] Refactor
- [ ] Other

### Tested Environments

- [ ] Development
- [ ] Staging
- [ ] Production

### Before submitting this PR:

- [x] Does your code build cleanly without any errors or warnings?
- [ ] Have you used auto closing keywords?
- [x] Have you added tests for new functionality?
- [ ] Have validated query efficiency for new database queries?
- [ ] Have documented new functionality in README or in comments?
- [x] Have you squashed all intermediate commits?
- [x] Is there a clear title that explains what the PR does?
- [x] Have you used intuitive function, variable and other naming?
- [ ] Have you requested security / privacy review if needed
- [x] Have you performed a self review of this PR?

### CODE REVIEW

File: wallet/controllers_v3_test.go [hunk 1]
```diff
@@ -260,3 +260,77 @@ func TestGetWalletV3(t *testing.T) {
...
+	// txHasDestination
+	var hasDestRows = sqlmock.NewRows([]string{"bool"}).AddRow(true)
+	mock.ExpectQuery("^select (.+)").WithArgs(idFrom).WillReturnRows(hasDestRows)
+
+	mock.ExpectExec("^UPDATE (.+)").WithArgs(idFrom, providerLinkingID, "bitflier").WillReturnResult(sqlmock.NewResult(1, 1))
+	mock.ExpectCommit()
+
+	ctx = context.WithValue(ctx, appctx.DatastoreCTXKey, datastore)
+	ctx = context.WithValue(ctx, appctx.RODatastoreCTXKey, roDatastore)
+	ctx = context.WithValue(ctx, appctx.ReputationClientCTXKey, mockReputation)
```
[the remark this answers is not in the document]
Comment (husobee) [reply wallet/controllers_v3_test.go:322]: I personally like the use of mocks primarily so that I don't need to stand up an entire environment to run these tests.  I also like the ability to inject errors into responses to test error conditions.  In these particular tests (wallet v3) there is no datastore actually setup, and these do not have the integration build flag.

File: wallet/service.go [hunk 2]
```diff
@@ -13,6 +13,7 @@ import (
 	errorutils "github.com/brave-intl/bat-go/utils/errors"
 	"github.com/brave-intl/bat-go/utils/handlers"
 	"github.com/brave-intl/bat-go/utils/logging"
+	"github.com/brave-intl/bat-go/utils/wallet"
```
[the remark this answers is not in the document]
Comment (husobee) [reply wallet/service.go]: fixed: https://github.com/brave-intl/bat-go/pull/656/commits/fd0bab3f847e3bc58b85713c12ceda4c187decec

File: wallet/controllers_v3_test.go [hunk 3]
```diff
@@ -260,3 +260,77 @@ func TestGetWalletV3(t *testing.T) {
...
+					"linkingInfo": "%s"
+				}`, tokenString)),
+		)
+		mockReputation = mockreputation.NewMockClient(mockCtrl)
+		handler        = wallet.LinkBitFlierDepositAccountV3(&wallet.Service{
+			Datastore: datastore,
+		})
+		w = httptest.NewRecorder()
+
+		walletClaimNamespace = uuid.Must(uuid.FromString("c39b298b-b625-42e9-a463-69c7726e5ddc"))
```
[the remark this answers is not in the document]
Comment (husobee) [reply wallet/controllers_v3_test.go]: fixed: https://github.com/brave-intl/bat-go/pull/656/commits/e6cdd68e04ec437a5d435fc86a7ee61155ecdb17

File: wallet/controllers_v3_test.go [hunk 4]
```diff
@@ -260,3 +260,77 @@ func TestGetWalletV3(t *testing.T) {
...
+				grantserver.Postgres{
+					DB: sqlx.NewDb(db, "postgres"),
+				},
+			})
+
+		// add the datastore to the context
+		ctx = context.WithValue(context.Background(), appctx.BitFlierJWTKeyCTXKey, string(secret))
+		r   = httptest.NewRequest(
+			"POST",
+			"/v3/wallet/bitflier/7def9cda-6a14-4fa1-be86-43da80e56d2c/claim",
```
[the remark this answers is not in the document]
Comment (husobee) [reply wallet/controllers_v3_test.go]: fixed: https://github.com/brave-intl/bat-go/pull/656/commits/e6cdd68e04ec437a5d435fc86a7ee61155ecdb17

File: wallet/controllers_v3_test.go [hunk 5]
```diff
@@ -260,3 +260,77 @@ func TestGetWalletV3(t *testing.T) {
 		must(t, "invalid response", fmt.Errorf("expected 201, got %d", resp.StatusCode))
 	}
 }
+
+func TestLinkBitFlierWalletV3(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+	var (
+		secret      = []byte("a jwt secret")
+		tokenString = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJhY2NvdW50X2hhc2giOiIxMjM0NTY3ODkwIiwiZGVwb3NpdF9pZCI6ImMzZmQ5YTBhLWM0ZDItNDdmYy1iZjdlLTlhZTJiYzk3MzNhMSIsImV4dGVybmFsX2FjY291bnRfaWQiOiIyIiwicmVxdWVzdF9pZCI6IjEiLCJ0aW1lc3RhbXAiOiIxOTcwLTAxLTAxVDAwOjAwOjAwWiJ9.1-PTPLYhMKJkscgoQAY8gQic0cPBIceD4eVTcJ9J-0k"
```
[the remark this answers is not in the document]
Comment (husobee) [reply wallet/controllers_v3_test.go]: fixed: https://github.com/brave-intl/bat-go/pull/656/commits/e6cdd68e04ec437a5d435fc86a7ee61155ecdb17

File: wallet/controllers_v3_test.go [hunk 6]
```diff
@@ -260,3 +260,77 @@ func TestGetWalletV3(t *testing.T) {
...
+	ctx = context.WithValue(ctx, appctx.RODatastoreCTXKey, roDatastore)
+	ctx = context.WithValue(ctx, appctx.ReputationClientCTXKey, mockReputation)
+
+	r = r.WithContext(ctx)
+
+	router := chi.NewRouter()
+	router.Post("/v3/wallet/bitflier/{paymentID}/claim", handlers.AppHandler(handler).ServeHTTP)
+	router.ServeHTTP(w, r)
+
+	if resp := w.Result(); resp.StatusCode != http.StatusOK {
```
[the remark this answers is not in the document]
Comment (husobee) [reply wallet/controllers_v3_test.go:354]: fixed: https://github.com/brave-intl/bat-go/pull/656/commits/e6cdd68e04ec437a5d435fc86a7ee61155ecdb17

File: wallet/inputs.go [hunk 7]
```diff
@@ -237,3 +242,91 @@ func (lbdar *LinkBraveDepositAccountRequest) HandleErrors(err error) *handlers.A
...
+	if err := tok.Claims([]byte(jwtKey), &base, &linkingInfo); err != nil {
+		return fmt.Errorf("failed to parse the linking info jwt token: %w", err)
+	}
+
+	blr.DepositID = linkingInfo.DepositID
+	blr.AccountHash = linkingInfo.AccountHash
+
+	if blr.AccountHash == "" || blr.DepositID == "" {
+		// failed to extract claims, or the token is invalid
+		return fmt.Errorf("failed to parse claims in LinkingInfo: %w", err)
```
[the remark this answers is not in the document]
Comment (husobee) [reply wallet/inputs.go]: fixed: https://github.com/brave-intl/bat-go/pull/656/commits/e6cdd68e04ec437a5d435fc86a7ee61155ecdb17

File: wallet/controllers_v3_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (husobee) [reply wallet/controllers_v3_test.go:346]: I suggest we chat about the pros and cons at some point of both methods.  I feel unit tests are better suited for mocking, and we could cut down on our test/build times as well as get more detailed edge case testing if we go with a mocking framework.  Integration tests are still required, but for happy path full flow integration tests imo, not just testing datastore model to postgres integrations.  Regardless, the v3 wallet testing as a whole is using mocking for integration tests, so I suggest we make a plan outside this PR to address if we want to change this.

**Review by maikelmclauflin — DISMISSED** [review]

**Review by maikelmclauflin — APPROVED** [review]
